ERCS/erc-7677.md

```diff
@@ -0,0 +1,296 @@
+  `0x${string}`, // Chain ID
+  Record<string, any> // Context
+];
+
+type GetPaymasterStubDataResult = Record<string, `0x${string}`> // Stub paymaster values
+```
+
+###### `pm_getPaymasterStubData` Example Parameters
+
+```json
+[
+  {
+    "sender": "0x...",
+    "nonce": "0x...",
+    "initCode": "0x",
+    "callData": "0x...",
+    "callGasLimit": "0x...",
+    "verificationGasLimit": "0x...",
+    "preVerificationGas": "0x...",
+    "maxFeePerGas": "0x...",
+    "maxPriorityFeePerGas": "0x...",
+    "paymasterAndData": "0x...",
+    "signature": "0x..."
+  },
+  "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789",
+  "0x2105",
+  {
+    // Illustrative context field. These should be defined by service providers.
+    "policyId": "962b252c-a726-4a37-8d86-333ce0a07299"
+  }
+]
+```
+
+###### `pm_getPaymasterStubData` Example Return Value
+
+Paymaster services MUST detect which entrypoint version the account is using and return the correct fields.
+
+For example, if using entrypoint v0.6:
+
+```json
+{
+  "paymasterAndData": "0x..."
+}
+```
+
+If using entrypoint v0.7:
+
+```json
+{
+  "paymaster": "0x...",
+  "paymasterData": "0x..."
+}
+```
+
+#### `pm_getPaymasterData`
+
+Returns values to be used in paymaster-related fields of a signed user operation. These are not stub values and will be used during user operation submission to a bundler. Similar to `pm_getPaymasterStubData`, accepts an unsigned user operation, entrypoint address, chain id, and a context object.
+
+This method also returns a `sponsor` object with `name` and `icon` fields. The `name` field is the name of the party sponsoring the transaction, and the `icon` field is a URI pointing to an image. Wallet developers MAY choose to display sponsor information to users. The `icon` string MUST be a data URI as defined in [RFC-2397]. The image SHOULD be a square with 96x96px minimum resolution. The image format is RECOMMENDED to be either lossless or vector based such as PNG, WebP or SVG to make the image easy to render on the wallet. Since SVG images can execute Javascript, wallets MUST render SVG images using the `<img>` tag to ensure no untrusted Javascript execution can occur.
+
+##### `pm_getPaymasterData` RPC Specification
+
+```typescript
+// [userOp, entryPoint, chainId, context]
+type GetPaymasterDataParams = [
+  // Below is specific to Entrypoint v0.6 but this API can be used with other entrypoint versions too
+  {
+    sender: `0x${string}`;
+    nonce: `0x${string}`;
+    initCode: `0x${string}`;
+    callData: `0x${string}`;
+    callGasLimit: `0x${string}`;
+    verificationGasLimit: `0x${string}`;
+    preVerificationGas: `0x${string}`;
+    maxFeePerGas: `0x${string}`;
+    maxPriorityFeePerGas: `0x${string}`;
+  }, // userOp
+  `0x${string}`, // Entrypoint
+  `0x${string}`, // Chain ID
+  Record<string, any> // Context
+];
+
+type GetPaymasterDataResult = {
+  sponsor?: { name: string; icon: string } // Sponsor info
+} & Record<string, `0x${string}`>; // Paymaster values
```

Comment:
I actually disagree with having these differ by entry point @lukasrosario https://github.com/ethereum/ERCs/pull/360/files/f9e15bfd482be08ac8a3147e8756a8c6a07c0e1d#r1559567260

Reply:
discussed with Wilson offline, these fields should be separate as currently written in the draft. ie `paymasterAndData` for v0.6 and `paymaster` / `paymasterData` for v0.7.

@pedrouid the intention is to make this compatible with potential future changes, where the fields might include something other than `paymasterAndData`, `paymaster`, `paymasterData`. what do you think about

```
type GetPaymasterDataResult = {
  sponsor?: { name: string; icon: string } // Sponsor info
  paymaster?: string // Paymaster address (for v0.7 or higher)
  paymasterData?: string;
  paymasterAndData?: string; // Paymaster Data
} & Record<string, `0x${string}`>;
```

makes the v0.6 and v0.7 fields explicit, but still allows for future changes.

Reply:
For `pm_getPaymasterStubData` on v0.7, I'd recommend having this return type (if you plan on having a structured schema) contain both `paymasterVerificationGasLimit` and `paymasterPostOpGasLimit`. The paymaster service can hardcode gas limits that it knows works with its contract. When provided through the `eth_estimateUserOperationGas` endpoint bundlers should skip estimating these values (but will validate once that they work), saving computation.

Many (most?) paymaster functions have pretty static gas consumption, so it doesn't make sense to estimate gas for them each time. Instead just use a hardcoded limit that is known to work for all inputs and onchain states.

Reply:
@dancoombs could you elaborate on the difficulties with estimating `paymasterVerificationGasLimit ` and `paymasterPostOpGasLimit`?

My initial impression is that it might be difficult to estimate although not impossible if using custom tracers?

Reply:
There are not difficulties estimating `paymasterVerificationGasLimit`.

My suggestion above is that, regardless of estimation difficulties, paymaster may want to signal to their users the gas values that should be used. This will allow the user/bundler to skip estimating paymaster gas limits.

Many accounts/paymasters have very static verification functions that don't need to be estimated for each user op, and a static limit value can be used that the paymaster or account developer knows will work each time, saving the bundler from performing the same computation each time. Potentially latency or even RPC cost savings would trickle down to the user/dapp. This suggestion allows the paymaster to communicate those limits.

Reply:
I'm pretty hesitant to now bring bundlers into this mix. 

@dancoombs is it likely / have you observed that paymaster gas values are insufficiently estimated by bundlers?

That was the motivation for going with the 2-step flow here (proposed by Dror). I.e. we routinely observed insufficient gas estimates from paymasters so we decided wallets / bundlers should estimate gas themselves. I could get behind your proposal if the issue is that bundlers similarly provide insufficient paymaster gas estimates. I am less convinced if this is just for cost savings.

Reply:
How is this bringing bundlers into the mix? Its suggesting that paymasters supply their own gas limits if they know what those gas limits need to be.

There are scenarios where bundlers/wallets don't have enough context to provide sufficient gas estimates, especially for postOp, and the paymaster needs to supply to get stable execution.

Reply:
We would need to specify that bundlers should skip paymaster gas estimation if paymaster gas values are supplied during an `eth_estimateUserOperationGas` call.

Reply:
Like I said though, if you're saying that it would be easy for bundlers to misestimate paymaster gas values-resulting in failed user op submission-then I can get on board with this. It sounds like you are saying bundlers misestimating the paymaster gas values could be common? Is that right?

Reply:
> We would need to specify that bundlers should skip paymaster gas estimation if paymaster gas values are supplied during an `eth_estimateUserOperationGas` call.

This is already the [case](https://github.com/eth-infinitism/account-abstraction/blob/develop/erc/ERCS/erc-4337.md#-eth_estimateuseroperationgas).

```
gas limits (and prices) parameters are optional, but are used if specified.
```

Reply:
Ah right. Ok, I can get on board with this. @hazim-j what do you think?

Reply:
I think I could get behind this. But it might be worth defining a concrete example for when a bundler's `paymasterPostOpGasLimit` can become unreliable and adding it to the rationale section. My understanding of this on a high level is if a state change happens between simulation and execution that causes two different code paths in `postOp` then I can see how that would cause an unreliable estimation.

Reply:
i've added the paymaster gas values to the `pm_getPaymasterStubData` response. going to resolve this for now to move things along and we can continue discussion on it if necessary